population.cpp

```diff
@@ -116,9 +116,10 @@ std::vector<double> create_rescaled_fitness_vec(std::vector<double> distance_fro
     return fitness_inds;
 }
 
-void change_nth_ind_net(population& p, size_t ind_index, network n)
+population change_nth_ind_net(population p, size_t ind_index, network n)
```

Comment:
This are not actually getter functions. Although I think this is a correct design choice, it goes a bit outside the scope of the issue. Also this might slow down the simulation consitently if these are used in actual code and not optimized away by the compiler. Also getting population by value and returning it by value causes problems in the program due to the templating.

Reply:
Ok so. Yes, I am aware this (and others below) are not getter functions! I know what a getter function is! But a lot of functions used to rely on getter functions that returned non-constant references, to directly change something obtained through a getter. For instance, this one is wrapped in `change_nth_ind_net` that takes simulation as an argument ; it then uses `s.get_pop()` to call this function. So if the getter is returning a constant ref now, we need to change the value at some point. We could get it by constant ref, but we'll still need to copy it and return it by value. And I think basically all the other things you commented on follow the same pattern/ problem. Do you have a good solution for this problem?

Reply:
Wow! A lot of exclamation marks. If this is irritating for you, you do not have to do it. This refactorings are not easy, and also not always crystal clear to me how they will roll out, so I can also take care of them.
I see your point though; but especially in this case, I think that copying by value will cause problems with the templating.
Why not make 
```c++
change_nth_ind_net(simulation, index, network)
``` 
a member function of simulation
```c++
simulation::change_nth_ind_net(index, network)
```
?
It indeed changes its internal state, so it could be a memebr function. 
I also would like to use a "by value" design eventually for most of our stuff, but in this case I suspect it is not really possible. 
I am also thinking we could make a more general purpose "accesser" function that does not overload the "getter" by accident. Similar to the `get_to_x` functions i created in `environment `for json saving. Although I think this is not really in the rule book of correct encapsulation practices.

So the solutions I see are 3:

1. Find a way in which copying by value does not mees up the templating of the network of individuals in population (difficult, I do not know enough to help you)
2.  Make the change functions in simulation memebrs and keep the change functions in population void return type
3. Create an accesser function `get_to_x` that returns by reference but that we do not risk to unadvertedly use to do read-only operations